core/src/main/c/share/byte_sink.cpp

```diff
@@ -0,0 +1,178 @@
+ * platforms, but this doesn't affect usage as we never input such values.
+ */
+static size_t next_pow2(size_t n)
+{
+    // See: https://jameshfisher.com/2018/03/30/round-up-power-2/
+    // In this portable code we use two different slightly different intrinsics
+    // for MSVC and others.
+    //  * __builtin_clz(l): counts the number of leading zeros.
+    //  * _BitScanReverse(64): counts the 0-based index of the highest bit.
+    // As such they need to be handled slightly differently.
+    const size_t prev = n - 1;
+
+#if defined(__GNUC__) || defined(__clang__)
+    const int n_leading_zeros = (size_t)__builtin_clzll(prev);
+    const size_t width = 64;
+    return ((size_t)1) << (width - n_leading_zeros);
+#else
+    unsigned long bit_index = 0;
+    _BitScanReverse64(&bit_index, prev);
+    return ((size_t)1) << (bit_index + 1);
+#endif
+}
+
+std::byte* questdb_byte_sink_book(questdb_byte_sink_t* sink, size_t min_len) {
+    const size_t curr_avail = sink->hi - sink->ptr;
+    if (curr_avail >= min_len) {
+        return sink->ptr;
+    }
+    const size_t curr_pos = sink->ptr - sink->lo;
+    const size_t curr_capacity = sink->hi - sink->lo;
+    const size_t add_req_capacity = min_len - curr_avail;
+    const size_t new_capacity = std::min(next_pow2(curr_capacity + add_req_capacity), max_alloc_size);
+
+    // TODO: Remove this check once `ByteSequence`'s `size` returns `long` instead of `int`.
+    // Cap allocation to 2GiB.
+    if ((new_capacity == max_alloc_size) && (new_capacity < (curr_capacity + add_req_capacity))) {
+        sink->overflow = true;
+        return NULL;
+    }
+
+    std::byte* const new_lo = (std::byte*) realloc(sink->lo, new_capacity);
+    if (new_lo == NULL) {
+        // NB: sink->lo is still valid here and will be freed later by `destroy`.
+        return NULL;
+    }
+    sink->lo = new_lo;
+    sink->hi = new_lo + new_capacity;
+    sink->ptr = new_lo + curr_pos;
+    return sink->ptr;
+}
+
+extern "C" {
+    JNIEXPORT jlong JNICALL Java_io_questdb_std_bytes_DirectByteSink_implCreate(
+            JNIEnv *env,
+            jclass cl,
+            jlong capacity);
+
+    JNIEXPORT void JNICALL Java_io_questdb_std_bytes_DirectByteSink_implDestroy(
+            JNIEnv *env,
+            jclass cl,
+            jlong impl);
+
+    JNIEXPORT jlong JNICALL Java_io_questdb_std_bytes_DirectByteSink_implBook(
+            JNIEnv *env,
+            jclass cl,
+            jlong impl,
+            jlong min_len);
+}
+
+JNIEXPORT jlong JNICALL Java_io_questdb_std_bytes_DirectByteSink_implCreate(
+        JNIEnv *env,
+        jclass cl,
+        jlong capacity) {
+    return (jlong) create(capacity);
+}
+
+JNIEXPORT void JNICALL Java_io_questdb_std_bytes_DirectByteSink_implDestroy(
+        JNIEnv *env,
+        jclass cl,
+        jlong impl) {
+    destroy((questdb_byte_sink_t*) impl);
+}
+
+JNIEXPORT void JNICALL JavaCritical_io_questdb_std_bytes_DirectByteSink_implDestroy(
```

Comment:
nit: `implCreate` doesn't have `JavaCritical_` variant. I thought we're getting rid of `JavaCritical_*`, don't we?

Reply:
I must have gotten distracted while I was removing them and only took out half of them. Thanks for spotting it. I've deleted the rest.